app/presenters/hyrax/av_file_set_presenter.rb

```diff
@@ -15,10 +15,79 @@
 # limitations under the License.
 # ---  END LICENSE_HEADER BLOCK  ---
 
+require 'avalon/manifest_range'
+
 module Hyrax
   class AVFileSetPresenter < Hyrax::FileSetPresenter
     include DisplaysContent
 
     Hyrax::MemberPresenterFactory.file_presenter_class = Hyrax::AVFileSetPresenter
+
+    attr_accessor :media_fragment
+
+    def range
+      structure_ng_xml.root.blank? ? simple_iiif_range : structure_to_iiif_range
+    end
+
+    private
+
+      def simple_iiif_range
+        # TODO: embed_title?
+        Avalon::ManifestRange.new(
+          label: { '@none'.to_sym => [title.first] },
+          items: [
+            clone.tap { |s| s.media_fragment = 't=0,' }
+          ]
+        )
+      end
+
+      def structure_to_iiif_range
+        div_to_iiif_range(structure_ng_xml.root)
+      end
+
+      def div_to_iiif_range(div_node)
+        items = div_node.children.select(&:element?).collect do |node|
+          if node.name == "Div"
+            div_to_iiif_range(node)
+          elsif node.name == "Span"
+            span_to_iiif_range(node)
+          end
+        end
+
+        # if a non-leaf node has no valid "Div" or "Span" children, then it would become empty range node containing no canvas
+        # raise an exception here as this error shall have been caught and handled by the parser and shall never happen here
+        raise Nokogiri::XML::SyntaxError, "Empty root or Div node: #{div_node[:label]}" if items.empty?
+
+        Avalon::ManifestRange.new(
+          label: { '@none'.to_sym => [div_node[:label]] },
+          items: items
+        )
+      end
+
+      def span_to_iiif_range(span_node)
+        Avalon::ManifestRange.new(
+          label: { '@none'.to_sym => [span_node[:label]] },
+          items: [
+            clone.tap do |s|
+              s.media_fragment = "t=#{parse_hour_min_sec(span_node[:begin])},#{parse_hour_min_sec(span_node[:end])}"
+            end
+          ]
+        )
+      end
+
+      def parse_hour_min_sec(s)
+        return nil if s.nil?
+        smh = s.split(':').reverse
+        (Float(smh[0]) rescue 0) + 60*(Float(smh[1]) rescue 0) + 3600*(Float(smh[2]) rescue 0)
```

Comment:
I think I previously found that having more logic and avoiding the rescue is much faster but I discovered that before this code was written.  I'll find the changes I made and make a PR.